mayastor/src/pool.rs

```diff
@@ -33,26 +35,66 @@ use spdk_sys::{
 };
 
 use crate::{
+    bdev::uring_util,
     core::Bdev,
     ffihelper::{cb_arg, done_cb},
     jsonrpc::{jsonrpc_register, Code, JsonRpcError, Result},
     replica::ReplicaIter,
 };
 
-/// Wrapper for create aio bdev C function
-fn create_base_bdev(file: &str, block_size: u32) -> Result<()> {
-    debug!("Creating aio bdev {} ...", file);
+/// Wrapper for create aio or uring bdev C function
+fn create_base_bdev(file: &str, block_size: u32, io_if: &str) -> Result<()> {
+    let (mut do_uring, must_uring) = match io_if {
+        "auto" | "" => (true, false),
+        "aio" => (false, false),
+        "uring" => (true, true),
+        _ => {
+            let err_str = "Invalid I/O interface: ".to_string() + io_if;
+            return Err(JsonRpcError::new(Code::InvalidParams, err_str));
+        }
+    };
+    if do_uring {
+        match uring_util::kernel_supports_io_uring() {
+            Ok(_) => {}
+            Err(e) => {
+                let err_str = "Uring not supported: ".to_string() + &e;
+                if must_uring {
+                    return Err(JsonRpcError::new(
+                        Code::InvalidParams,
+                        err_str,
+                    ));
+                } else {
+                    warn!("{}, falling back to aio for bdev {}", err_str, file);
+                    do_uring = false;
+                }
+            }
+        }
+    }
+    let bdev_type = if !do_uring {
```

Comment:
Not sure what this tuple is needed for...

Reply:
It's just to get the subsequent log entries looking correct.